### PR TITLE
Add support for CudaHalfTensor, CudaDoubleTensor training

### PR DESCRIPTION
**Pull request by gchanan**
The -precision flag now supports single/double/half for specifying tensor precision with either single or multi gpu (DataParallel) mode.

Note that this does not add support for the pre-trained models
in classify, extract-features.

### CODE REVIEW

**Comment by gchanan** [comment]
This is work continued from https://github.com/facebook/fb.resnet.torch/pull/134, but now supports multi-GPU as well (and all the pre-reqs have been committed to cutorch, cunn (for DataParallelTable), etc).

**Comment by colesbury** [comment]
This looks great. Is it ready to be merged?

**Comment by gchanan** [comment]
I've been using this for a couple weeks now without issue, so I think it's ready to go.

**Comment by thadpasce16** [comment]
How much training accuracy loss do you see when going from single to half precision?

**Comment by gchanan** [comment]
It depends on if you use fp16 or fp32 compute (cutorch.hasFastInstructions)

Additional top-1 error compared to fp32 storage / fp32 compute:
fp16 storage / fp16 compute: +17%
fp16 storage / fp32 compute: +4%
